lib/pbench/server/api/resources/query_apis/search_api.py

```diff
@@ -0,0 +1,126 @@
+        # use the unconverted strings passed by the caller rather than the
+        # adjusted and normalized datetime objects for this.
+        start_arg = f"{start:%Y-%m}"
+        end_arg = f"{end:%Y-%m}"
+
+        self.logger.info(
+            "Search query for user {}, prefix {}: ({} - {}) on query: {}",
+            user,
+            self.prefix,
+            start,
+            end,
+            query,
+        )
+
+        uri_fragment = self._gen_month_range("run", start, end)
+        return {
+            "path": f"/{uri_fragment}/_search",
+            "kwargs": {
+                "json": {
+                    "query": {
+                        "bool": {
+                            "filter": [
+                                {"term": self._get_user_term(user)},
+                                {
+                                    "range": {
+                                        "@timestamp": {"gte": start_arg, "lte": end_arg}
+                                    }
+                                },
+                            ],
+                            "must": {
+                                "query_string": {
+                                    "query": f"*{query}*",
+                                    "analyze_wildcard": "true",
+                                }
+                            },
+                        }
+                    },
+                    "sort": [
+                        {"@timestamp": {"order": "desc", "unmapped_type": "boolean"}}
+                    ],
+                    "_source": {"include": selected_fields},
+                },
+                "params": {"ignore_unavailable": "true"},
+            },
+        }
+
+    def postprocess(self, es_json: Dict[AnyStr, Any]) -> Dict[AnyStr, Any]:
+        """
+        """
+        results = []
+        # If there are no matches for the user, query, and time range,
+        # return the empty list rather than failing.
+        try:
+            count = es_json["hits"]["total"]["value"]
+            if int(count) == 0:
+                self.logger.warning("No data returned by Elasticsearch")
+                return jsonify(result)
+        except KeyError as e:
+            raise PostprocessError(
+                f"Can't find Elasticsearch match data {e} in {es_json!r}"
+            )
+        except ValueError as e:
+            raise PostprocessError(f"Elasticsearch hit count {count!r} value: {e}")
+        hits = es_json["hits"]["hits"]
+        for hit in hits:
+            s = flatten_json(hit["_source"])
+            s["id"] = hit["_id"]
+            results.append(s)
+        # construct response object
+        return jsonify(results)
+
+
+# Utility function for flattening the json
+def flatten_json(es_json, prefix="", out={}):
+    if type(es_json) is dict:
+        for key in es_json:
+            while prefix[-2:-1] == ".":
+                prefix = prefix[:-1]
```

Comment:
Why not just `split` on `"."` like the dashboard does? And this only looks for a "." at the end??

Reply:
The effect has a lot of explicit allowances for missing values (represented as `'undefined'`) that you're missing; you may be taking `KeyError`s here that'll short-circuit your processing erroneously.

Reply:
Yeah we need to remove the trailing dots. Can you give me an example, I tested the function on the `_source` of elasticsearch hit. 
The flatten function would take the json like following:
```
{
      "@timestamp": "2021-07-12T22:44:19.562354",
      "@metadata": {
          "file-date": "2021-07-12T23:03:00.810683",
          "file-name": "/srv/pbench/archive/fs-version-001/dhcp31-171.perf.lab.eng.bos.redhat.com/pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T22.44.19.tar.xz",
          "file-size": 3347324,
          "md5": "1c25e9f5b5dfc1ffb732931bf3899878",
          "toc-prefix": "pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T22.44.19",
          "pbench-agent-version": "0.69.8-1g90314ee7c",
          "controller_dir": "dhcp31-171.perf.lab.eng.bos.redhat.com"
      },
      "@generated-by": "b01699a7a9ccfaed84cdec883c7ce6dd",
      "authorization": {
          "owner": "3",
          "access": "private"
      },
      "run": {
          "controller": "dhcp31-171.perf.lab.eng.bos.redhat.com",
          "name": "pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T22.44.19",
          "script": "pbench-user-benchmark",
          "config": "npalaska-dhcp31-171",
          "date": "2021-07-12T22:44:19",
          "iterations": "1-default",
          "toolsgroup": "default",
          "start": "2021-07-12T22:44:19.562354",
          "end": "2021-07-12T22:44:32.358148",
          "id": "1c25e9f5b5dfc1ffb732931bf3899878"
      },
      "sosreports": [
          {
              "name": "pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T22.44.19/sysinfo/end/dhcp31-171/sosreport-dhcp31-171-pbench-dhcp31-171.perf.lab.eng.bos.redhat.com-2021-07-12-kpqchzn.tar.xz",
              "md5": "cda0d8e931b46e57fcc736165559fd74",
              "hostname-f": "dhcp31-171.perf.lab.eng.bos.redhat.com",
              "hostname-s": "dhcp31-171",
              "inet": [
                  {
                      "ifname": "lo",
                      "ipaddr": "127.0.0.1"
                  },
                  {
                      "ifname": "ens3",
                      "ipaddr": "10.16.31.171"
                  }
              ],
              "inet6": [
                  {
                      "ifname": "lo",
                      "ipaddr": "::1"
                  },
                  {
                      "ifname": "ens3",
                      "ipaddr": "2620:52:0:101c:5054:ff:fe5e:5a4a"
                  },
                  {
                      "ifname": "ens3",
                      "ipaddr": "fe80::5054:ff:fe5e:5a4a"
                  }
              ]
          }
      ],
      "host_tools_info": [
          {
              "hostname": "dhcp31-171",
              "hostname-f": "dhcp31-171.perf.lab.eng.bos.redhat.com",
              "tools": {
                  "iostat": "--interval=3",
                  "mpstat": "--interval=3",
                  "perf": "--record-opts='record -a --freq=100'",
                  "pidstat": "--interval=30",
                  "proc-interrupts": "--interval=3",
                  "proc-vmstat": "--interval=3",
                  "sar": "--interval=3",
                  "turbostat": "--interval=3"
              }
          }
      ]
  }
```
and output it like this: 
```
"@timestamp":"2021-07-12T22:44:19.562354",
   "@metadata.file-date":"2021-07-12T23:03:00.810683",
   "@metadata.file-name":"/srv/pbench/archive/fs-version-001/dhcp31-171.perf.lab.eng.bos.redhat.com/pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T22.44.19.tar.xz",
   "@metadata.file-size":3347324,
   "@metadata.md5":"1c25e9f5b5dfc1ffb732931bf3899878",
   "@metadata.toc-prefix":"pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T22.44.19",
   "@metadata.pbench-agent-version":"0.69.8-1g90314ee7c",
   "@metadata.controller_dir":"dhcp31-171.perf.lab.eng.bos.redhat.com",
   "@generated-by":"b01699a7a9ccfaed84cdec883c7ce6dd",
   "authorization.owner":"3",
   "authorization.access":"private",
   "run.controller":"dhcp31-171.perf.lab.eng.bos.redhat.com",
   "run.name":"pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T22.44.19",
   "run.script":"pbench-user-benchmark",
   "run.config":"npalaska-dhcp31-171",
   "run.date":"2021-07-12T22:44:19",
   "run.iterations":"1-default",
   "run.toolsgroup":"default",
   "run.start":"2021-07-12T22:44:19.562354",
   "run.end":"2021-07-12T22:44:32.358148",
   "run.id":"1c25e9f5b5dfc1ffb732931bf3899878",
   "sosreports.name":"pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T22.44.19/sysinfo/end/dhcp31-171/sosreport-dhcp31-171-pbench-dhcp31-171.perf.lab.eng.bos.redhat.com-2021-07-12-kpqchzn.tar.xz",
   "sosreports.md5":"cda0d8e931b46e57fcc736165559fd74",
   "sosreports.hostname-f":"dhcp31-171.perf.lab.eng.bos.redhat.com",
   "sosreports.hostname-s":"dhcp31-171",
   "sosreports.inet.ifname":"ens3",
   "sosreports.inet.ipaddr":"10.16.31.171",
   "sosreports.inet6.ifname":"ens3",
   "sosreports.inet6.ipaddr":"fe80::5054:ff:fe5e:5a4a",
   "host_tools_info.hostname":"dhcp31-171",
   "host_tools_info.hostname-f":"dhcp31-171.perf.lab.eng.bos.redhat.com",
   "host_tools_info.tools.iostat":"--interval=3",
   "host_tools_info.tools.mpstat":"--interval=3",
   "host_tools_info.tools.perf":"--record-opts='record -a --freq=100'",
   "host_tools_info.tools.pidstat":"--interval=30",
   "host_tools_info.tools.proc-interrupts":"--interval=3",
   "host_tools_info.tools.proc-vmstat":"--interval=3",
   "host_tools_info.tools.sar":"--interval=3",
   "host_tools_info.tools.turbostat":"--interval=3"
```

If the value is undefined something like `"key": { "sub_key1": "1", "sub_key2": None}` it would get flatten like this `key.sub_key2: None` is that what you were asking?

Reply:
Just stating here that this flattening algorithm has changed a bit and it would produce the following output now:

```
{
   "@timestamp":"2021-07-12T22:44:19.562354",
   "@metadata.file-date":"2021-07-12T23:03:00.810683",
   "@metadata.file-name":"/srv/pbench/archive/fs-version-001/dhcp31-171.perf.lab.eng.bos.redhat.com/pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T22.44.19.tar.xz",
   "@metadata.file-size":3347324,
   "@metadata.md5":"1c25e9f5b5dfc1ffb732931bf3899878",
   "@metadata.toc-prefix":"pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T22.44.19",
   "@metadata.pbench-agent-version":"0.69.8-1g90314ee7c",
   "@metadata.controller_dir":"dhcp31-171.perf.lab.eng.bos.redhat.com",
   "@generated-by":"b01699a7a9ccfaed84cdec883c7ce6dd",
   "authorization.owner":"3",
   "authorization.access":"private",
   "run.controller":"dhcp31-171.perf.lab.eng.bos.redhat.com",
   "run.name":"pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T22.44.19",
   "run.script":"pbench-user-benchmark",
   "run.config":"npalaska-dhcp31-171",
   "run.date":"2021-07-12T22:44:19",
   "run.iterations":"1-default",
   "run.toolsgroup":"default",
   "run.start":"2021-07-12T22:44:19.562354",
   "run.end":"2021-07-12T22:44:32.358148",
   "run.id":"1c25e9f5b5dfc1ffb732931bf3899878",
   "sosreports.name":"pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T22.44.19/sysinfo/end/dhcp31-171/sosreport-dhcp31-171-pbench-dhcp31-171.perf.lab.eng.bos.redhat.com-2021-07-12-kpqchzn.tar.xz",
   "sosreports.md5":"cda0d8e931b46e57fcc736165559fd74",
   "sosreports.hostname-f":"dhcp31-171.perf.lab.eng.bos.redhat.com",
   "sosreports.hostname-s":"dhcp31-171",
   "sosreports.inet.ifname":[
      "lo",
      "ens3"
   ],
   "sosreports.inet.ipaddr":[
      "127.0.0.1",
      "10.16.31.171"
   ],
   "sosreports.inet6.ifname":[
      "lo",
      "ens3",
      "ens3"
   ],
   "sosreports.inet6.ipaddr":[
      "::1",
      "2620:52:0:101c:5054:ff:fe5e:5a4a",
      "fe80::5054:ff:fe5e:5a4a"
   ],
   "host_tools_info.hostname":"dhcp31-171",
   "host_tools_info.hostname-f":"dhcp31-171.perf.lab.eng.bos.redhat.com",
   "host_tools_info.tools.iostat":"--interval=3",
   "host_tools_info.tools.mpstat":"--interval=3",
   "host_tools_info.tools.perf":"--record-opts='record -a --freq=100'",
   "host_tools_info.tools.pidstat":"--interval=30",
   "host_tools_info.tools.proc-interrupts":"--interval=3",
   "host_tools_info.tools.proc-vmstat":"--interval=3",
   "host_tools_info.tools.sar":"--interval=3",
   "host_tools_info.tools.turbostat":"--interval=3"
}
```